Add Scylla Cloud serverless support for CQLSH

Lorak-mmk opened the pull request:
This PR adds serverless support for CQLSH.


Fixes https://github.com/scylladb/scylla-tools-java/issues/317

avelanarius commented:
Please add "Fixes scylladb/scylla-tools-java#317" to this PR.

fruch reviewed (approved):
LGTM

fruch commented:
@Lorak-mmk @avelanarius 
a new release for python-driver, in currently building:
https://github.com/scylladb/python-driver/actions/runs/3471804467

Lorak-mmk commented:
Python driver version is changed, so this is PR is now ready

Lorak-mmk commented on pylib/requirements.txt:
```diff
@@ -3,7 +3,7 @@
 futures
 six
--e git+https://github.com/datastax/python-driver.git@cassandra-test#egg=cassandra-driver
+scylla-driver==3.25.8
```
I'll modify `build-resolver.xml` accordingly. Where in Scylla repo are the tests using cqlsh? I think there are some in dtests - but I don't know how to tell scylla-ci which branch of dtests to use

fruch replied:
Use the master branch in dtest

fruch replied:
The tests you are looking for are here:

https://github.com/scylladb/scylla-dtest/blob/next/cqlsh_tests/cqlsh_tests.py

Lorak-mmk commented:
V3 (?):
- Fixed 2 bugs preventing usage without `--cloudconf` (@fruch I see you made a PR to https://github.com/scylladb/scylla-cqlsh/pull/6 - it should probably be updated to reflect that).
- Modifications to Ant files - I checked and this PR now works both with Ant and with requirements.txt and can connect to both Cloud and normal Clusters.
- Better commit history / messages.

TODO:
Testing - most likely by utilizing Scylla CI as Israel suggested.

fruch commented:
> V3 (?):
> 
> * Fixed 2 bugs preventing usage without `--cloudconf` (@fruch I see you made a PR to [Add Scylla Cloud serverless support scylla-cqlsh#6](https://github.com/scylladb/scylla-cqlsh/pull/6) - it should probably be updated to reflect that).

I've tested it base on ccm sni_proxy support, so I've fixed this issue (please verify it in my PR)

> * Modifications to Ant files - I checked and this PR now works both with Ant and with requirements.txt and can connect to both Cloud and normal Clusters.
> * Better commit history / messages.
> 
> TODO: Testing - most likely by utilizing Scylla CI as Israel suggested.

Lorak-mmk commented:
> > V3 (?):
> > 
> > * Fixed 2 bugs preventing usage without `--cloudconf` (@fruch I see you made a PR to [Add Scylla Cloud serverless support scylla-cqlsh#6](https://github.com/scylladb/scylla-cqlsh/pull/6) - it should probably be updated to reflect that).
> 
> I've tested it base on ccm sni_proxy support, so I've fixed this issue (please verify it in my PR)

I left one comment, and also changed my PR slightly according to your changes.

> 
> > * Modifications to Ant files - I checked and this PR now works both with Ant and with requirements.txt and can connect to both Cloud and normal Clusters.
> > * Better commit history / messages.
> > 
> > TODO: Testing - most likely by utilizing Scylla CI as Israel suggested.

Lorak-mmk commented:
Rebased on master - previous version didn't contain commit from Avi that fixes CI error (https://github.com/scylladb/scylla-tools-java/commit/caf754f2434a55ad04e9dd147b8d97c62b8d8427 )

Lorak-mmk commented:
dtests passed with current version: https://github.com/scylladb/scylladb/pull/12025

fruch reviewed (approved):
LGTM

nyh commented:
> dtests passed with current version: [scylladb/scylladb#12025](https://github.com/scylladb/scylladb/pull/12025)

It seems to sproadically fail various dtests now - see https://github.com/scylladb/scylla-dtest/issues/2923.
We may need to revert this patch...

nyh commented:
The failures seem to all have the same problem: They use the "USE" command and a later requests that specifies only a table name claims it doesn't have a default keyspace set.

My unproven theory is that the Scylla driver might be using multiple connections (why? maybe some shard-awareness thing?) and forgetting that cqlsh's "do_use()" needs to run the USE statement on all connections, not just one.

fruch commented:
> The failures seem to all have the same problem: They use the "USE" command and a later requests that specifies only a table name claims it doesn't have a default keyspace set.
> 
> My unproven theory is that the Scylla driver might be using multiple connections (why? maybe some shard-awareness thing?) and forgetting that cqlsh's "do_use()" needs to run the USE statement on all connections, not just one.

I could be something similar, but it's not in the control of cqlsh, it's on the driver end
it's something like 

Session -> Pools (one per host) -> Connections (on per shared)

when connecting:
```
session = cluster.connect()
# or
cluster.connect(wait_for_all_pools=True)
```

it would wait for all hosts to be connected, but the extra shard aware connection are happening at the background, 

hence they might not have received the `RESULT_KIND_SET_KEYSPACE = 0x0003` response from a `USE keyspace` CQL command, since they weren't yet opened

nyh commented:
@fruch opened https://github.com/scylladb/python-driver/issues/187 reporting the bug in the Scylla python driver which we started experiencing after this PR switched us to the Scylla python driver.

avikivity commented:
I queued a revert.
